client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java

```diff
@@ -27,29 +27,36 @@
 import org.elasticsearch.action.update.UpdateResponse;
 import org.elasticsearch.client.Requests;
 import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.common.collect.MapBuilder;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.rest.BaseRestHandler;
 import org.elasticsearch.rest.BytesRestResponse;
 import org.elasticsearch.rest.RestChannel;
-import org.elasticsearch.rest.RestController;
 import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.RestRequest.Method;
 import org.elasticsearch.rest.RestResponse;
 import org.elasticsearch.rest.action.RestBuilderListener;
 
 import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
 
+import static java.util.Arrays.asList;
+import static java.util.Collections.unmodifiableList;
 import static org.elasticsearch.rest.RestRequest.Method.POST;
 import static org.elasticsearch.rest.RestRequest.Method.PUT;
 import static org.elasticsearch.rest.RestStatus.OK;
 
 public class RestNoopBulkAction extends BaseRestHandler {
 
-    public RestNoopBulkAction(RestController controller) {
-        controller.registerHandler(POST, "/_noop_bulk", this);
-        controller.registerHandler(PUT, "/_noop_bulk", this);
-        controller.registerHandler(POST, "/{index}/_noop_bulk", this);
-        controller.registerHandler(PUT, "/{index}/_noop_bulk", this);
+    @Override
+    public Map<String, List<Method>> handledMethodsAndPaths() {
+        return Collections.unmodifiableMap(MapBuilder.<String, List<Method>>newMapBuilder()
+            .put("/_noop_bulk", unmodifiableList(asList(POST, PUT)))
+            .put("/{index}/_noop_bulk", unmodifiableList(asList(POST, PUT)))
+            .map());
```

Comment:
Or:

```diff
diff --git a/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java b/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java
index d6c0ef736e3..b0d17af1846 100644
--- a/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java
+++ b/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java
@@ -27,7 +27,6 @@ import org.elasticsearch.action.support.ActiveShardCount;
 import org.elasticsearch.action.update.UpdateResponse;
 import org.elasticsearch.client.Requests;
 import org.elasticsearch.client.node.NodeClient;
-import org.elasticsearch.common.collect.MapBuilder;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.rest.BaseRestHandler;
@@ -39,12 +38,9 @@ import org.elasticsearch.rest.RestResponse;
 import org.elasticsearch.rest.action.RestBuilderListener;
 
 import java.io.IOException;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
-import static java.util.Arrays.asList;
-import static java.util.Collections.unmodifiableList;
 import static org.elasticsearch.rest.RestRequest.Method.POST;
 import static org.elasticsearch.rest.RestRequest.Method.PUT;
 import static org.elasticsearch.rest.RestStatus.OK;
@@ -53,10 +49,9 @@ public class RestNoopBulkAction extends BaseRestHandler {
 
     @Override
     public Map<String, List<Method>> handledMethodsAndPaths() {
-        return Collections.unmodifiableMap(MapBuilder.<String, List<Method>>newMapBuilder()
-            .put("/_noop_bulk", unmodifiableList(asList(POST, PUT)))
-            .put("/{index}/_noop_bulk", unmodifiableList(asList(POST, PUT)))
-            .map());
+        return Map.of(
+            "/_noop_bulk", List.of(POST, PUT),
+            "/{index}/_noop_bulk", List.of(POST, PUT));
     }
 
     @Override
```

I [had been working](https://github.com/elastic/elasticsearch/commit/f48ddd53fd66cfef3032a17023cd9e2f88702af9#diff-adff2a5e0bb2632b7661a17ae93a267d) to remove uses of `MapBuilder` in master, in light of the new JDK APIs for map construction.

Reply:
And I wonder if we should take this farther, and instead of using a simple datatype we formalize the notion of a route and make that part of the API:

```diff
diff --git a/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java b/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java
index d6c0ef736e3..1e70bc9f15b 100644
--- a/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java
+++ b/client/client-benchmark-noop-api-plugin/src/main/java/org/elasticsearch/plugin/noop/action/bulk/RestNoopBulkAction.java
@@ -27,7 +27,6 @@ import org.elasticsearch.action.support.ActiveShardCount;
 import org.elasticsearch.action.update.UpdateResponse;
 import org.elasticsearch.client.Requests;
 import org.elasticsearch.client.node.NodeClient;
-import org.elasticsearch.common.collect.MapBuilder;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.rest.BaseRestHandler;
@@ -37,14 +36,12 @@ import org.elasticsearch.rest.RestRequest;
 import org.elasticsearch.rest.RestRequest.Method;
 import org.elasticsearch.rest.RestResponse;
 import org.elasticsearch.rest.action.RestBuilderListener;
+import org.elasticsearch.rest.action.Route;
 
 import java.io.IOException;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 
-import static java.util.Arrays.asList;
-import static java.util.Collections.unmodifiableList;
 import static org.elasticsearch.rest.RestRequest.Method.POST;
 import static org.elasticsearch.rest.RestRequest.Method.PUT;
 import static org.elasticsearch.rest.RestStatus.OK;
@@ -52,11 +49,10 @@ import static org.elasticsearch.rest.RestStatus.OK;
 public class RestNoopBulkAction extends BaseRestHandler {
 
     @Override
-    public Map<String, List<Method>> handledMethodsAndPaths() {
-        return Collections.unmodifiableMap(MapBuilder.<String, List<Method>>newMapBuilder()
-            .put("/_noop_bulk", unmodifiableList(asList(POST, PUT)))
-            .put("/{index}/_noop_bulk", unmodifiableList(asList(POST, PUT)))
-            .map());
+    public List<Route> handledMethodsAndPaths() {
+        return List.of(
+            new Route("/_noop_bulk", List.of(POST, PUT)),
+            new Route("/{index}/_noop_bulk", List.of(POST, PUT)));
     }
 
     @Override
diff --git a/server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java b/server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java
index 26ba8aa27fb..66d65b0b70e 100644
--- a/server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java
+++ b/server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java
@@ -28,6 +28,7 @@ import org.elasticsearch.common.settings.Setting;
 import org.elasticsearch.common.settings.Setting.Property;
 import org.elasticsearch.plugins.ActionPlugin;
 import org.elasticsearch.rest.RestRequest.Method;
+import org.elasticsearch.rest.action.Route;
 import org.elasticsearch.rest.action.admin.cluster.RestNodesUsageAction;
 
 import java.io.IOException;
@@ -74,7 +75,7 @@ public abstract class BaseRestHandler implements RestHandler {
      * {@inheritDoc}
      */
     @Override
-    public abstract Map<String, List<Method>> handledMethodsAndPaths();
+    public abstract List<Route> handledMethodsAndPaths();
 
     @Override
     public final void handleRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
diff --git a/server/src/main/java/org/elasticsearch/rest/RestController.java b/server/src/main/java/org/elasticsearch/rest/RestController.java
index 239f89c5c20..9bf52ced99d 100644
--- a/server/src/main/java/org/elasticsearch/rest/RestController.java
+++ b/server/src/main/java/org/elasticsearch/rest/RestController.java
@@ -160,8 +160,8 @@ public class RestController implements HttpServerTransport.Dispatcher {
      * and {@code path} combinations.
      */
     public void registerHandler(RestHandler restHandler) {
-        restHandler.handledMethodsAndPaths().forEach((path, methods) ->
-            methods.forEach(method -> registerHandler(method, path, restHandler)));
+        restHandler.handledMethodsAndPaths().forEach(route ->
+            route.methods().forEach(method -> registerHandler(method, route.path(), restHandler)));
         restHandler.deprecatedHandledMethodsAndPaths().forEach(api -> api.getMethods().forEach(method ->
             registerAsDeprecatedHandler(method, api.getPath(), restHandler, api.getDeprecationMessage(), api.getLogger())));
         restHandler.replacedMethodsAndPaths().forEach(api -> registerWithDeprecatedHandler(api.getMethod(), api.getPath(), restHandler,
diff --git a/server/src/main/java/org/elasticsearch/rest/RestHandler.java b/server/src/main/java/org/elasticsearch/rest/RestHandler.java
index 5f0862ed80e..6504656ee3c 100644
--- a/server/src/main/java/org/elasticsearch/rest/RestHandler.java
+++ b/server/src/main/java/org/elasticsearch/rest/RestHandler.java
@@ -23,6 +23,7 @@ import org.elasticsearch.client.node.NodeClient;
 import org.elasticsearch.common.logging.DeprecationLogger;
 import org.elasticsearch.common.xcontent.XContent;
 import org.elasticsearch.rest.RestRequest.Method;
+import org.elasticsearch.rest.action.Route;
 
 import java.util.Collections;
 import java.util.List;
@@ -69,8 +70,8 @@ public interface RestHandler {
     /**
      * The map of {@code path} to {@code methods} that this RestHandler is responsible for handling.
      */
-    default Map<String, List<Method>> handledMethodsAndPaths() {
-        return Collections.emptyMap();
+    default List<Route> handledMethodsAndPaths() {
+        return List.of();
     }
 
     /**
diff --git a/server/src/main/java/org/elasticsearch/rest/action/Route.java b/server/src/main/java/org/elasticsearch/rest/action/Route.java
index d42b4fb851a..05e38fdce74 100644
--- a/server/src/main/java/org/elasticsearch/rest/action/Route.java
+++ b/server/src/main/java/org/elasticsearch/rest/action/Route.java
@@ -1,4 +1,26 @@
 package org.elasticsearch.rest.action;
 
+import org.elasticsearch.rest.RestRequest;
+
+import java.util.List;
+
 public class Route {
+
+    private final String path;
+
+    public String path() {
+        return path;
+    }
+
+    private List<RestRequest.Method> methods;
+
+    public List<RestRequest.Method> methods() {
+        return methods;
+    }
+
+    public Route(final String path, final List<RestRequest.Method> methods) {
+        this.path = path;
+        this.methods = methods;
+    }
+
 }
```

This is close to what you have done with deprecated routes, but I think we should take it the whole way and cover all routes. Then a `DeprecatedRoute` is a `Route` with some additional fields, etc.